cmd/zoekt-dynamic-indexserver/main.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2016 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This program manages a zoekt dynamic indexing deployment:
+// * listens to indexing commands
+// * reindexes specified repositories
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+)
+
+func loggedRun(cmd *exec.Cmd) (out, err []byte) {
+	outBuf := &bytes.Buffer{}
+	errBuf := &bytes.Buffer{}
+	cmd.Stdout = outBuf
+	cmd.Stderr = errBuf
+
+	log.Printf("run %v", cmd.Args)
+	if err := cmd.Run(); err != nil {
+		log.Printf("command %s failed: %v\nOUT: %s\nERR: %s",
+			cmd.Args, err, outBuf.String(), errBuf.String())
+	}
+
+	return outBuf.Bytes(), errBuf.Bytes()
+}
+
+type Options struct {
+	indexTimeout time.Duration
+	dataDir      string
+	indexDir     string
+	repoDir      string
+	listen       string
+}
+
+func (o *Options) createMissingDirectories() {
+	for _, s := range []string{o.dataDir, o.indexDir, o.repoDir} {
+		if _, err := os.Stat(s); err == nil {
+			continue
+		}
+
+		if err := os.MkdirAll(s, 0o755); err != nil {
+			log.Fatalf("MkdirAll %s: %v", s, err)
+		}
+	}
+}
+
+type indexRequest struct {
+	CloneURL string // TODO: Decide if tokens can be in the URL or if we should pass separately
+	RepoID   uint32
+}
+
+// This function is declared as var so that we can stub it in test
+var executeCmd = func(ctx context.Context, name string, arg ...string) {
+	cmd := exec.CommandContext(ctx, name, arg...)
+	cmd.Stdin = &bytes.Buffer{}
+	loggedRun(cmd)
+}
+
+func indexRepository(opts Options, req indexRequest, w http.ResponseWriter) {
+	ctx, cancel := context.WithTimeout(context.Background(), opts.indexTimeout)
```

Comment:
I'm not sure why you are passing w in here. You don't write anything to it unless filepath.Abs fails (unlikely), and you don't respect the requests timeout (via r.GetContext()).

Reply:
I've removed `w` from `indexRepository`, thanks!

> and you don't respect the requests timeout (via r.GetContext()).

Do you mean that we should take `r.Context().Deadline()`? Right now `(deadline time.Time, ok bool)` returns `ok == false`, which means that we don't have one and will use `opts.indexTimeout`

Reply:
It probably makes sense to not respect the ctx to be honest, since indexing a repository can take longer than what is reasonable for a request.

However, you can use `CommandContext` to respect the ctx when running a command. Then if the user cancels the request it will cancel the index. But yeah, it likely makes sense to just ignore canceled requests to ensure you still index the repo.

Reply:
> However, you can use CommandContext to respect the ctx when running a command. Then if the user cancels the request it will cancel the index. But yeah, it likely makes sense to just ignore canceled requests to ensure you still index the repo.

I think ignoring cancelled requests makes more sense, thanks!